src/core/lender/test/mock/assessor.sol

```diff
@@ -1,14 +1,57 @@
 pragma solidity >=0.4.24;
 
-contract AssessorMock {
+import "ds-math/math.sol";
 
-    uint public assetReturn; function setAssetReturn(uint assetAmount_) public {assetReturn=assetAmount_;}
-    
-    uint public wad;
-    address public tranche;
+contract AssessorMock is DSMath {
 
-    function getAssetValueFor(address tranche_) public returns (uint) {
-        tranche = tranche_;
-        return assetReturn;
+    uint256 constant ONE = 10 ** 27;
+
+    address public senior;
+    address public junior;
+
+    mapping (bytes32 => uint) calls;
+    mapping (bytes32 => uint) returnValues;
+
+
+    function setReturn(bytes32 name, uint returnValue) public {
+        returnValues[name] = returnValue;
+    }
+
+    function call(bytes32 name) internal returns (uint) {
+        calls[name]++;
+        return returnValues[name];
+    }
+
+    function calcTokenPrice () public returns (uint) {
+        return call("tokenPrice");
+    }
+
+    function calcAssetValue(address tranche) public returns (uint) {
+        uint trancheReserve;
+        if (tranche == junior) {
+            trancheReserve = returnValues["jBalance"];
+        }
+        trancheReserve = returnValues["sBalance"];
+        uint poolValue = returnValues["pileDebt"];
+        if (tranche == junior) {
+            return calcJuniorAssetValue(poolValue, trancheReserve, seniorDebt());
+        }
+        return calcSeniorAssetValue(poolValue, trancheReserve, seniorDebt(), juniorReserve());
+    }
+
+    // Tranche.assets (Junior) = (Pool.value + Tranche.reserve - Senior.debt) > 0 && (Pool.value - Tranche.reserve - Senior.debt) || 0
+    function calcJuniorAssetValue(uint poolValue, uint trancheReserve, uint seniorDebt) internal returns (uint) {
+        int assetValue = int(poolValue + trancheReserve - seniorDebt);
+        return (assetValue > 0) ? uint(assetValue) : 0;
+    }
+    // Tranche.assets (Senior) = (Tranche.debt < (Pool.value + Junior.reserve)) && (Senior.debt + Tranche.reserve) || (Pool.value + Junior.reserve + Tranche.reserve)
+    function calcSeniorAssetValue(uint poolValue, uint trancheReserve, uint trancheDebt, uint juniorReserve) internal returns (uint) {
+        return ((poolValue + juniorReserve) >= trancheDebt) ? (trancheDebt + trancheReserve) : (poolValue + juniorReserve + trancheReserve);
+    }
+    function juniorReserve() internal returns (uint) {
+        return call("juniorReserve");
+    }
+    function seniorDebt() internal returns (uint) {
```

Comment:
Is there actually a reason this needs to exist on the assessor at all? (I know I'm reviewing the mock - not the impl) but I'm curious!

Reply:
we need it in the junior calculation for the case where no senior tranche exists
```javascript
  function seniorDebt() internal returns (uint) {
        return (senior != address(0x0)) ? SeniorTrancheLike(senior).debt() : 0;
    }
```

we could try to call `SeniorTrancheLike(0x00...000).debt()` and see if it fails. maybe it would return just
a 0 because the contract doesn't exists.

I prefer the ckeck, I think